### ISSUE TITLE
CPU #3: profile and gate fused dequant+dot

### ISSUE THREAD

**Issue by Nando2392** [Nando2392/qx-mini-c#26]
## Hypothesis
CPU priority #3: investigate and, only if justified, implement minimal opt-in fused dequant+dot to avoid materializing 256-float temporaries before dot products.

## Causal inventory so far
Static scan plus real runtime probe identify the material seam as the F32 final `lm_head` path:

| Route | Location | Tensor/type | Temporary | Frequency estimate | Modes | Notes |
|---|---|---|---:|---:|---|---|
| final head F32 | `src/qx_format.c:3435` in `qx_compute_real_final_head` | `output.weight`, Q6_K (`ggml_type=14`) | `float weights[256]` per block | `vocab_size * (dims/256)` = `151936 * 8` = `1,215,488` block decodes per final-head call; `311,165,?` temporary floats (~1.16 GiB writes) | F32; buffered/mmap | Real probe confirms `lm_head_kernel=dequant_f32`, full vocab `151936`, dims `2048`, argmax `358`, bench decode phase ~8.69s for generate=1.
| final head Q8_K | `src/qx_format.c:3420` -> `qx_compute_real_final_head_q8_k` | `output.weight`, Q6_K | no F32 block temp expected | same rows/blocks but fused Q6_K x Q8_K | `q8_k_compat`; buffered/mmap | Real probe confirms `lm_head_kernel=q6_k_q8_k`, argmax `358`, decode phase ~2.431s for generate=1.
| projection fallback | `src/qx_format.c:3999` in `qx_projection_matvec_fill_mode` | projection tensors; non-Q8_K fallback | `float vals[256]` | rows * blocks_per_row | F32 fallback | Existing `q8_k_projection` path already calls Q5/Q6/IQ4 fused Q8_K kernels; IQ4_XS prefix already specialized.
| packed expert F32 | `src/qx_format.c:4501` in `qx_packed_expert_matvec_mode` | MoE gate/up/down packed experts | `float weights_block[256]` | selected experts * rows * blocks_per_row | F32 fallback | `q8_k_compat` workspace path already avoids temp for types 17/18/21/22/23.
| embedding/probes/helpers | multiple decode/block probe paths | embeddings/probes | `float[256]` | low/probe-only | both/probe | Not first target unless profiling disproves final head.

Real runtime probes executed before opening this issue:

```text
F32 buffered ephemeral generate=1 final-head:
selected=[358]
bench elapsed=16.812s; prefill=8.122s; decode=8.690s
lm_head_kernel=dequant_f32; logits_computed=151936; input_dims=2048; lm_head_ggml_type=14

q8_k_compat buffered ephemeral generate=1 final-head:
selected=[358]
bench elapsed=4.535s; prefill=2.104s; decode=2.431s
lm_head_kernel=q6_k_q8_k; cache_readback_ok=true
```

These probes establish that the final-head path is executed in real runtime and that the F32 path still materializes Q6_K blocks before dot. They do **not** prove a new F32 fused kernel will improve wall-clock; that requires counters/benchmark in this issue.

## Scope
Allowed:
- add opt-in profiling/counters for dequant+dot materialization;
- quantify blocks, temporary floats/bytes, fused/fallback calls by activation/backend/kernel policy and phase;
- add RED->GREEN tests for counters and, if warranted, a scalar Q6_K dequant+dot fused helper for final head only;
- keep fallback exact and fail-closed.

Rejected in this issue:
- threading, SIMD/AVX2/FMA, CUDA, expert cache, QXF format changes, KV2, speculative decoding, prefill GEMM, general kernel refactor, default promotion.

## TDD plan
1. RED: benchmark validator rejects missing `dequant_dot_profile`.
2. RED: CLI rejects invalid kernel/profile policy values and defaults to baseline.
3. RED: final-head F32 profile reports Q6_K temp block counts/floats/bytes for baseline.
4. RED: fused opt-in path reports lower temporary bytes and nonzero fused calls while selected tokens/checksums remain exact vs baseline.
5. RED: fallback/unsupported type remains baseline and reports fallback calls.
6. GREEN minimal implementation only after RED tests exist.

## Benchmark/evidence gate
Write new evidence, not overwriting Issues #24/#25:

`wiki/evidence/issue-26-fused-dequant-dot-baseline.json`

Minimum matrix:
- kernel policy: `baseline|fused`
- activation: `f32|q8_k_compat`
- backend: `buffered|mmap`
- 1 warmup, >=3 measured reps
- fixed model/tokenizer/prompt/layers/ctx/KV/seed/build/provenance

Correctness gates:
- exact outputs within each activation/backend/kernel policy where exactness is expected;
- prompt token IDs, selected/final tokens, final residual/norm/logits/KV checksums, `cache_readback_ok`;
- do not require F32 == q8_k_compat.

Performance/causal gates:
- fused reduces temporary bytes or dequant calls in the target path;
- no material unexplained RSS or wall-clock regression;
- if profiling refutes material benefit, publish evidence and do not promote/default.

**Comment by Nando2392** [comment]
COMPLETED in 9c2d3727cdbc55a3e3a54bfd62659271e09d17da.

Evidence:
- CI: https://github.com/Nando2392/qx-mini-c/actions/runs/32509580751 — SUCCESS (windows-msvc 2m17s)
- Legal-diff-SHA256: b4009d0cea40181c02529edf379620c5c70b259312145449b5d6c497fd576975
- Evidence JSON: wiki/evidence/issue-26-fused-dequant-dot-baseline.json
- MSVC C17 build: PASS (qxfit.c
qx_main.c
Generating Code...
qx_format.c
qx_gguf.c
qx_tokenizer.c
qx_qxf_main.c
Generating Code...
qx_format.c
qx_gguf.c
qx_tokenizer.c
state_loop_replay_api_contract.c
Generating Code...
qx_format.c
qx_gguf.c
qx_tokenizer.c
qxf_mmap_api_contract.c
Generating Code...)
- Focused tests: [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                                               [100%][0m
[32m[32m[1m10 passed[0m[32m in 0.05s[0m[0m → 10 passed
- QXF tests: [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                                                   [100%][0m
[32m[32m[1m6 passed[0m[32m in 0.73s[0m[0m → 6 passed
- Q8_K local non-WDAC subset: [32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m.[0m[32m                                                                    [100%][0m
[32m[32m[1m5 passed[0m, [33m4 deselected[0m[32m in 0.11s[0m[0m → 5 passed, 4 deselected; full local oracle exe path is blocked by WinError 4551 Device Guard on this host, CI Windows passed exact SHA.
- Smoke: {"status": "pass", "probe": "state_loop", "delta_source": "rope_gqa_attention", "layers_run": 4, "kv_appends": 4, "cache_readback_ok": true, "softmax_sum": 1, "golden_probe": "rope_gqa_golden", "tokenizer_qxt2": true, "tokenizer_ids": [7], "real_qkv_golden": true, "real_48_layer_state": true, "real_layers_run": 96, "real_carry_links": 47, "real_final_head": true, "real_argmax_token": 1124, "real_autoregressive_sequence": [1124, 50853]} → PASS, real_48_layer_state true, final sequence [1124, 50853]
- Wiki lint: PASS
- Auto Research: 20260821T173608Z ready_for_review, staged payload present, modelReferencesVerified=true, canUseForProductChange=true
- Independent reviews: security/correctness/numerics PASS; benchmark/spec/provenance PASS for digest b4009d0cea40181c02529edf379620c5c70b259312145449b5d6c497fd576975

Delivered:
- Opt-in  for final-head Q6_K dequant+dot profiling path.
-  counters for final-head Q6_K blocks, temporary floats/bytes, fused calls, fallback calls.
- Direct shell 2×2×2 kernel/activation/backend matrix: f32 buffered+mmap reduce temporary bytes/calls; q8_k_compat remains compatibility/control; no default promotion.

**Comment by Nando2392** [comment]
COMPLETED clean evidence summary for Issue #26.

Commit: 9c2d3727cdbc55a3e3a54bfd62659271e09d17da
CI: https://github.com/Nando2392/qx-mini-c/actions/runs/32509580751 — SUCCESS
Legal-diff-SHA256: b4009d0cea40181c02529edf379620c5c70b259312145449b5d6c497fd576975
Evidence JSON: wiki/evidence/issue-26-fused-dequant-dot-baseline.json

Delivered:
- Opt-in `--kernel-policy baseline|fused` for final-head Q6_K dequant+dot profiling path.
- `--dequant-profile` counters for final-head Q6_K blocks, temporary floats/bytes, fused calls, fallback calls.
- Direct shell 2×2×2 kernel/activation/backend matrix: f32 buffered+mmap reduce temporary bytes/calls; q8_k_compat remains compatibility/control; no default promotion.
- WDAC-safe smoke_check fallback: Windows `build/qxqxf.exe` remains primary; WSL fallback is used only after WinError 4551 Device Guard blocks Python CreateProcess locally.

Local gates:
- MSVC C17 build: PASS (`cmd.exe /c build_msvc.bat`).
- Focused benchmark tests: `python -m pytest tests/test_q8k_perf_experiment.py -q` → 10 passed.
- QXF tests: `python -m pytest tests/test_qxf.py -q` → 6 passed.
- Q8_K non-WDAC subset: `python -m pytest tests/test_q8k_activation.py -q -k "not test_q8_k_activation_bytes_match_pinned_ggml_oracle"` → 5 passed, 4 deselected. Full local oracle exe path is blocked on this host by WinError 4551 Device Guard; CI Windows passed exact SHA.
- Smoke: `python smoke_check.py` → PASS, real_48_layer_state=true, final sequence [1124, 50853].
- Wiki lint: PASS.
- Py compile: PASS.
- `git diff --cached --check`: PASS before commit.

Release gates:
- Hermes Auto Research run 20260821T173608Z: ready_for_review, stagedPayloadPresent=true, stagedDiffSha256 matched, modelReferencesVerified=true, canUseForProductChange=true.
- Independent digest reviews: security/correctness/numerics PASS; benchmark/spec/provenance PASS for digest b4009d0cea40181c02529edf379620c5c70b259312145449b5d6c497fd576975.
